### DISCUSSION TITLE
sync problem with thunderbird and google calendar

### DISCUSSION THREAD

**Discussion by ERRVV1N** [kewisch/gdata-provider#497] [Support]
I'm using TB 102.2.2 (64-bit) on Ubuntu 20.04.5 with gdata-provider 102.0.1 and the syncronization with Google Calendar hangs and TB is unusable.

**Comment by nneul** [comment]
I've been seeing this same problem, going into add-ons and choosing to disable the provider, (click past the unsubscribe dialog) and then immediately re-enabling provider seems to clear it up.

**Reply by kir0ul** [reply]
That fixed the issue for me :+1:

**Reply by kir0ul** [reply]
But after some time the issue came back... :sweat:

**Reply by henry88g** [reply]
This didn't work for me, but restarting TB did.  Fix is working after a while - will see how long it holds for.

**Comment by msjohns1** [comment]
I'm seeing a related problem. Synchronizing upstream changes does not function, from either the automatic sync or manually selecting to synchronize. It goes through the process, and does not cause TB to hang, but it also doesn't actually pull down any changes. If I fully close TB and reopen, that seems to trigger a successful sync. I haven't tried @nneul's suggestion, but I wonder if that would similarly trigger the sync. End result is the calendar state is unreliable.

TB 102.2.2 (64-bit) on Mint 20.3, Provider 102.0.0

**Reply by mackintg** [reply]
I have exact same versions as @msjohns1 and getting the exact same symptoms.

**Reply by msjohns1** [reply]
@mackintg I just tried @nneul's suggestion -- disable Provider, hit "Cancel" on the unsubscribe prompt, and immediately re-enable -- and then I was able to successfully create an entry directly in Google and a manual sync bring it into TB. So I recommend giving that a try; hopefully it resolves for you as well.

**Reply by mackintg** [reply]
@msjohns1 yes, I also tried @nneul's suggestion and it temporarily fixed the problem... but only for ~1 day and then it was back to no-sync of changes made on the gmail calendar outside of TB. I already downloaded the latest version 102.0.3 that is meant to resolve this problem (https://github.com/kewisch/gdata-provider/discussions/519#discussioncomment-4147395) but no improvement is apparent. 

My intuition is that a fix should be reasonably easy to resolve since most people can get TB calendars to sync by using @nneul's method or by just restarting TB. This means that there is a known state/config where TB syncs correctly and efficiently, so I hope that the developer is reviewing these comments and can create the same state whenever a calendar sync is required.

**Comment by xjonx1** [comment]
The app has to be updated. The developer has been contacted. The problem is the app needs google's new OAuth 2.0.

**Comment by ERRVV1N** [comment]
I was able to find out that it is due to Thunderbird and not the addon. The addon works very well in TB version 102.4.2 (64-bit/Ubuntu 20.04.5 with gdata-provider 102.0.1). I haven't been able to test all the features yet, but it seems to be working very well.

**Reply by msjohns1** [reply]
Thanks for the shout. I upgraded TB (but still using provider 102.0.0 since that's what the extension manager says is current) and my testing was successful. So hopefully this resolves things.

**Comment by xjonx1** [comment]
There was an issue with the add-on and is now fixed by the developer. Details here: https://github.com/kewisch/gdata-provider/discussions/519#discussioncomment-4147395

**Reply by madtom1999** [reply]
Not working on 102.4.2 on 64 bit Xubuntu.

**Comment by nneul** [comment]
The updated plugin fixes the authentication issue, however I don't believe it fixes the sync problem in THIS issue - I saw a 'stopped syncing' instance yesterday AFTER I upgraded to 102.0.3. Will be keeping an eye on it to see if that problem reoccurs.

**Reply by xjonx1** [reply]
I had the plug-in self-remove itself from Thunderbird once but removing the old calendar and re-adding it seems to have fixed that problem. I merely mention that to say, did you remove(unsubscribe) from your calendar then add it back in? That may help, or it may not.

**Comment by mackintg** [comment]
Interesting that @msjohns1 found this problem was fixed by upgrading TB, because I am using TB 102.5.0 and the very latest gdata-provider 102.0.3 from github ([here](https://github.com/kewisch/gdata-provider/discussions/519#discussioncomment-4147395))   , but I find that the calendar sync problem persists. So I am in the same situation as @nneul ... the gdata-provider update to 102.0.3 fixes the authentication problem, but not the sync problem that is the focus of this thread.

Various tricks, like i) disabling the add-on and adding it back, ii) unsubscribing from the calendar and then re-subscribing to rebuilding the calendar from scratch, iii) restarting TB ... these all seem to clear the problem for a while. But my experience is that the no-sync problem eventually returns when an existing calendar entry is modified outside of TB, and that modification will not sync into TB until I do one of the things mentioned above.

As I posted earlier, my hope is that a fix should be forthcoming since most people can get TB calendars to sync by using @nneul's method or by just restarting TB. This means that there is a known state/config where gdata-provider will sync correctly and efficiently, so I hope that the developer is reviewing these comments and can create that same state whenever a calendar sync is required.

**Comment by dellskin** [comment]
I am on TB 102.5.1 64-bit and Provider 102.0.3. I think this was working(maybe, can't remember) in TB 102.5.0, but not working now (my TB updated yesterday). Turning off and on the Provider does not work, it only loads and fixes data as of that moment. It does not fix the problem on an ongoing basis.

**Comment by MaxMotovilov** [comment]
102.5.1/64 + 102.0.3. Calendars sync when TB starts up and, I think, never again.

**Comment by mackintg** [comment]
I don't think this is going to be fixed any time soon. The work-around that I have settled on takes advantage of the fact that restarting TB fixes the calendar sync problem **for a period of time** (but unfortunately the exact circumstances that lead to the calendar sync failing again are still unknown). I created a system-idle trigger to run a script when I am not using my computer, which restarts TB if it is active (e.g. using cron in Linux, or Task Manager in Windows). Restarting TB every couple of hours (if idle for, say, 30 minutes) seems to keep the problem under control for me while I wait for a real fix!

**Comment by kboeckx** [comment]
Same issue here, sync it only performed when starting TB or disabling/enabling the add-on. Please fix as this issue is there since a few mounths.

**Comment by Numiah** [comment]
As others I am still struggling with this issue.
Using TB 102.7.2 and GC 102.0.3
Solution for me is disabling the extension and using the google.calendar through a webbrowser.

**Comment by carlosabalde** [comment]
Anything we could do to help troubleshooting this? I'm not familiar with TB plugin development, but providing some basic guidance on the data to be collected that could be useful to diagnose whatever is going on here, I'd be very happy to help.

**Comment by carlosabalde** [comment]
I've been trying to do some troubleshooting on my own (zero knowledge on TB plugin development). I used the TB Config Editor to enable `calendar.debug.log` and `calendar.debug.log.verbose` (not sure if that changes anything, but it's documented in the repo), and then I used the Error Console (`Ctrl` + `Shift` + `J`).

Every time I clicked 'Calendars > Synchronize', I saw some reasonable logging in the console:

```
10:17:22.771
GEThttps://www.googleapis.com/calendar/v3/calendars/addressbook#contacts@group.v.calendar.google.com/events?maxResults=1000&showDeleted=true&syncToken=EDITED
[HTTP/2 200 OK 294ms]

10:17:22.772
GEThttps://www.googleapis.com/calendar/v3/calendars/EDITED@group.calendar.google.com/events?maxResults=1000&showDeleted=true&syncToken=EDITED
[HTTP/2 200 OK 158ms]

10:17:22.773
GEThttps://www.googleapis.com/calendar/v3/calendars/EDITED/events?maxResults=1000&showDeleted=true&syncToken=EDITED
[HTTP/2 200 OK 141ms]

10:17:22.838
GEThttps://www.googleapis.com/calendar/v3/calendars/EDITED@import.calendar.google.com/events?maxResults=1000&showDeleted=true&syncToken=EDITED
[HTTP/2 200 OK 150ms]

10:17:22.839
GEThttps://www.googleapis.com/calendar/v3/calendars/EDITED@group.v.calendar.google.com/events?maxResults=1000&showDeleted=true&syncToken=EDITED
[HTTP/2 200 OK 166ms]

10:17:22.840
GEThttps://www.googleapis.com/tasks/v1/lists/EDITED/tasks?maxResults=1000&showCompleted=true&showHidden=true&updatedMin=2023-03-01T09:16:10Z&showDeleted=true
```

Sometimes I clicked the button and nothing was logged, but I assume some logic exists to ignore updates if they are requested too frequently. However, at some point (yesterday it took several hours), clicking the button stops logging anything at all. I've been looking for previous log messages, but nothing seems related with this issue. Of course, as commented several times here, disabling-enabling the add-on restores the normal behavior.

I'd love to help here, but I'd need some guidance in order to know what I should adjust to get additional logging and hopefully get some clues :)

**Reply by carlosabalde** [reply]
Btw, I forgot to mention: TB 102.7.1 (Ubuntu Jammy) + Provider for Google Calendar 102.0.3.

**Reply by carlosabalde** [reply]
I've just learned that restarting TB is needed in order for `calendar.debug.log` and `calendar.debug.log.verbose` to take effect. Already done and now waiting for the issue to repeat...

**Reply by carlosabalde** [reply]
Ok, same behavior as yesterday: at some point things break and clicking 'Calendars > Synchronize' stops working. Nothing is logged when the additional debug logging is in place, but I think I have a clue. Last messages logged by the add-on before sync stops working are these:

```
18:33:18.602 Calendar: [calCachedCalendar] Doing changelog based sync for calendar googleapi://EDITED@gmail.com/?calendar=addressbook%23contacts%40group.v.calendar.google.com calCachedCalendar.js:301
18:33:18.602 Calendar: [calGoogleCalendar] Skipping refresh since user is idle gdataCalendar.jsm:758
18:33:18.613 Calendar: [calCachedCalendar] Doing changelog based sync for calendar googleapi://EDITED@gmail.com/?calendar=EDITED%40group.calendar.google.com calCachedCalendar.js:301
18:33:18.614 Calendar: [calGoogleCalendar] Skipping refresh since user is idle gdataCalendar.jsm:758
18:33:18.631 Calendar: [calCachedCalendar] Doing changelog based sync for calendar googleapi://EDITED@gmail.com/?calendar=EDITED%40gmail.com calCachedCalendar.js:301
18:33:18.631 Calendar: [calGoogleCalendar] Skipping refresh since user is idle gdataCalendar.jsm:758
18:33:18.641 Calendar: [calCachedCalendar] Doing changelog based sync for calendar googleapi://EDITED@gmail.com/?calendar=EDITED%40import.calendar.google.com calCachedCalendar.js:301
18:33:18.642 Calendar: [calGoogleCalendar] Skipping refresh since user is idle gdataCalendar.jsm:758
18:33:19.015 Calendar: [calCachedCalendar] Doing changelog based sync for calendar googleapi://EDITED@gmail.com/?calendar=EDITED%40group.v.calendar.google.com calCachedCalendar.js:301
18:33:19.015 Calendar: [calGoogleCalendar] Skipping refresh since user is idle gdataCalendar.jsm:758
18:33:19.265 Calendar: [calCachedCalendar] Doing changelog based sync for calendar googleapi://EDITED@gmail.com/?tasks=EDITED calCachedCalendar.js:301
18:33:19.266 Calendar: [calGoogleCalendar] Skipping refresh since user is idle
```

That's logged in https://github.com/kewisch/gdata-provider/blob/main/src/background/calendar.js#L375. Does this help?

**Reply by carlosabalde** [reply]
Ping @kewisch :)

**Reply by carlosabalde** [reply]
One more day testing, same behavior: once the `Skipping refresh since user is idle` messages are logged, the add-on stops working.

**Reply by Numiah** [reply]
Perhaps you should try the edit button to save us followers from a notification every time you add something...

**Reply by carlosabalde** [reply]
OMG, such a delightful, helpful, friendly & collaborative attitude! I'd love to return the favor and suggest you click the unsubscribe button, but instead I'll keep waiting for your contribution to the problem. Have a nice day you too! ;)

**Reply by carlosabalde** [reply]
Based on my previous observations, I've cloned the repo, made a minor adjustment to the implementation, rebuilt and installed the modified add-on. So far the sync issue seems to have been fixed, although I'd like to let this run for a few days before jumping to any conclusions.

In any case, for the sake of those who may be interested, the modification is as simple as skipping [the logic to bypass sync when the user is idle](https://github.com/kewisch/gdata-provider/blob/v102.0.3/src/legacy/modules/gdataCalendar.jsm#L757-L761). Anybody here with a deeper understanding than me of the add-on's internals who knows the negative implications of this modification / can suggest a less hacky solution? @kewisch I'm looking at you :) It'd also be great if others experiencing the same issue could confirm if this fixes the problem for them too.

**Reply by carlosabalde** [reply]
Here it is the built add-on including the described fix, for those who want to give it a try: [gdata-provider.xpi.gz](https://github.com/kewisch/gdata-provider/files/10883548/gdata-provider.xpi.gz). For me, so far, no more sync issues.

**Reply by nneul** [reply]
What commit/branch/tag did you base your changes on? Maybe I'm just out of date/sync. I'll take a look at your build. (I was against oauth2, trying now against legacy).

**Reply by carlosabalde** [reply]
Tag `v102.0.3`, which is the latest release.

**Comment by nneul** [comment]
Looking at that chunk of code, I think the issue is that it's tripping on the "idle" - and then never re-enabling the handler here:

https://github.com/kewisch/gdata-provider/blob/375ad8cd5aed5d1f1891c8bc48adf6dd778df24f/src/legacy/modules/gdataCalendar.jsm#L901

              this.replayChangesOn(aListener);

A more "correct" fix might be just to skip all the "processing" pieces in that function, and jump straight to that last function that resets the callback.

**Reply by carlosabalde** [reply]
Makes sense, but my JavaScript knowledge is basically nonexistent :) In any case, as long as we can confirm the problem is in there, writing the right fix should be simple. Did the ugly hack suggested by me fix the problem for you too?

**Reply by nneul** [reply]
Giving the fix a try now and will let you know.

**Reply by nneul** [reply]
This may be something with what I built - but I'm getting the google login every 30 minutes now - doesn't appear to be "sticking".

**Reply by carlosabalde** [reply]
I've using the modified add-on since yesterday and that didn't happen. You can give a try to the once I attached in the previous thread.

**Reply by nneul** [reply]
Since rebuilding against 'legacy' branch instead of oauth I haven't seen any issues with the workaround and it seems stable.

**Comment by jachoo** [comment]
I've managed to workaround the bug without rebuilding or reinstalling. Just run in the console (Tools -> Developer Tools -> Debug Addons -> Provide for Google Calendar -> Console) following line:

`await messenger.storage.local.set({"settings.idleTime": 0})`

After executing this line, one must restart Thunderbird. From now, the calendar sync should work correctly.

**Reply by kboeckx** [reply]
This is a good wordarround ! I just tested it and it works fine !

**Reply by henry88g** [reply]
Silly question.  I'm not familiar with the debugging console.  Mine shows double chevron (>>) at the start of the line.  Do I include the single chevron in the line --   **> await messenger.storage.local.set({"settings.idleTime": 0})**  or not?

**Reply by hellsbelz** [reply]
I'm not familiiar, too. Including a ">" at the beginning results in a syntax error, while omitting the ">" returns an "<- undefined", whatever this means here. 

Anyway, I will observe the add-ons behaviour the next days.

**Reply by rag2** [reply]
It seems to work for me, too. It was applied a couple of days ago. I'll report again—or try to—should the fixed Thunderbird (version 102.8.0 (64-bit) on Ubuntu 18.04) display the _fails-to-sync_ syndrome again.

It has been a **very** annoying bug.

**Reply by JayBird707** [reply]
Thanks work for me also!

**Reply by retiredfeline** [reply]
I confirm that this works for me too. Thunderbird 115.1.0 on 64-bit Linux, gdata-provider 115.0.2. Now synchronisation is reliably bidirectional instead of Thunderbird -> Google reliable. Can this be fixed in future releases?

**Reply by matt-buckley** [reply]
Worked for me as well with the more recent issues being reported!

**Reply by atscomms** [reply]
Worked for me, so far!
Thanks!

**Reply by otaolafranc** [reply]
Hello,
I just found myself with the same issue, and I can not find 'Tools -> Developer Tools -> Debug Addons -> Provide for Google Calendar -> Console'
version 115.9.0 (64-bit)
any ideas/recommendations?

**Reply by paeddi** [reply]
With the new UI you should find the console under Addons, than the gear symbol (manage extensions), debug Add-Ons, and then inspect (I hope it is called like that in english UI...) for the Provider for Google Calendar.
but I'm not sure if that's still the solution... ;-)

**Reply by otaolafranc** [reply]
Hello @paeddi thanks for your answer, I managed to do it (I think), but the problem reminds for me sadly

**Comment by kewisch** [comment]
Anyone who is working around this issue with the settings.idleTime hack, I need help to debug/fix this in https://github.com/kewisch/gdata-provider/issues/498 .
